channeld/channel.c

```diff
@@ -2156,21 +2126,72 @@ static void send_shutdown_complete(struct peer *peer)
 	}
 
 	/* Now we can tell master shutdown is complete. */
-	daemon_conn_send(&peer->master,
-			 take(towire_channel_shutdown_complete(peer,
-							       &peer->pcs.cs)));
-	daemon_conn_send_fd(&peer->master, PEER_FD);
-	daemon_conn_send_fd(&peer->master, GOSSIP_FD);
-
-	if (!daemon_conn_sync_flush(&peer->master))
-		status_failed(WIRE_CHANNEL_INTERNAL_ERROR, "Flushing master");
+	wire_sync_write(MASTER_FD,
+			take(towire_channel_shutdown_complete(peer,
+							      &peer->pcs.cs)));
+	fdpass_send(MASTER_FD, PEER_FD);
+	fdpass_send(MASTER_FD, GOSSIP_FD);
+	close(MASTER_FD);
+}
+
+static bool process_reqs(struct peer *peer)
+{
+	const u8 *msg;
+	bool changed = false;
+
+	/* In case we've deferred, process packet backlog. */
+	while ((msg = msg_dequeue(&peer->from_master)) != NULL) {
+		status_trace("Now dealing with deferred %s",
+			     channel_wire_type_name(fromwire_peektype(msg)));
+		req_in(peer, msg);
+		changed = true;
+	}
+
+	return changed;
+}
+
+static struct peer *peer;
+
+/* If this becomes a common pattern, we could make it a helper in common/ */
+static int poll_with_masterfd(struct pollfd *fds, nfds_t nfds, int timeout)
```

Comment:
Am I correct in assuming this hooks into `io` to add checking the master fd for new messages? Interesting trick :wink:

Reply:
Yeah, it allows for almost arbitrary manipulation.